### PR TITLE
Add CAPI status/conditions to deployment status

### LINKED ISSUE #219: Update Deployment status based on CAPI objects condition

### PR DESCRIPTION
**Pull request by kylewuolle**
Resolves #219. This PR adds a new function that fetches the CAPI status / conditions of the cluster and places them in the deployment status/conditions. To ensure the conditions are updated the controller now requeues until the CAPI cluster conditions are all true.

### CODE REVIEW

File: internal/controller/deployment_controller.go [hunk 1]
```diff
@@ -87,6 +89,63 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+		return true, fmt.Errorf("failed to create client configuration: %v", err)
+	}
+
+	resourceId := schema.GroupVersionResource{
+		Group:    "cluster.x-k8s.io",
+		Version:  "v1beta1",
+		Resource: "clusters",
+	}
+
+	list, err := dc.Resource(resourceId).Namespace(deployment.Namespace).Get(ctx, deployment.Name, metav1.GetOptions{})
```
Comment (Kshatrix) [inline internal/controller/deployment_controller.go:104]: I think it's more reliable to use helm labels instead of a specific name as naming patterns are not enforced. I think it's fine to assume that we expect exactly one cluster object for now.

**Review by Kshatrix — DISMISSED** [review]

File: internal/controller/deployment_controller.go [hunk 2]
```diff
@@ -87,6 +89,63 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+	if apierrors.IsNotFound(err) {
+		l.Info("Cluster not found, ignoring since object must be deleted or not yet created")
+		return true, nil
+	}
+
+	if err != nil {
+		return true, fmt.Errorf("failed to get cluster information for deployment: %s : %v", deployment.Name, err)
```
Comment (squizzi) [inline internal/controller/deployment_controller.go:111]:
```suggestion
		return true, fmt.Errorf("failed to get cluster information for deployment: %s: %w", deployment.Name, err)
```

File: internal/controller/deployment_controller.go [hunk 3]
```diff
@@ -87,6 +89,63 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+	}
+
+	requeue := true
+	if statusField, ok := list.Object["status"]; ok {
+		if statusMap, ok := statusField.(map[string]interface{}); ok {
+			if conditionsField, ok := statusMap["conditions"]; ok {
+				bytes, err := json.Marshal(conditionsField)
+				if err != nil {
+					return true, fmt.Errorf("failed to get deserialize cluster information for deployment: %s : %v",
```
Comment (squizzi) [inline internal/controller/deployment_controller.go:120]:
```suggestion
					return true, fmt.Errorf("failed to get deserialize cluster information for deployment: %s: %w",
```

File: internal/controller/deployment_controller.go [hunk 4]
```diff
@@ -87,6 +89,63 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+		l.Info("Cluster not found, ignoring since object must be deleted or not yet created")
+		return true, nil
+	}
+
+	if err != nil {
+		return true, fmt.Errorf("failed to get cluster information for deployment: %s : %v", deployment.Name, err)
+	}
+
+	requeue := true
+	if statusField, ok := list.Object["status"]; ok {
```
Comment (squizzi) [inline internal/controller/deployment_controller.go:115]: You don't need to do this, the API returns `unstructured.Unstructured` types and there's an entire library to help you make these conversions and get this content much much easier:

```go
	conditions, found, err := unstructured.NestedMap(list.Object, "status", "conditions")
	if err != nil {
		return true, fmt.Errorf("failed to get cluster information for deployment: %s: %w", deployment.Name, err)
	}
	if !found {
		return true, fmt.Errorf("failed to get cluster information for deployment: %s: status.conditions not found", deployment.Name)
	}

	var conditionsList []metav1.Condition

	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(conditions, &conditionsList); err != nil {
		return true, fmt.Errorf("failed to convert unstructured conditions to []metav1.Condition for deployment: %s: %w", deployment.Name, err)
	}

	allConditionsComplete := true
	for _, condition := range conditionsList {
		if condition.Status != "True" {
			allConditionsComplete = false
		}

		if condition.Reason == "" && condition.Status == "True" {
			condition.Reason = "Succeeded"
		}
		apimeta.SetStatusCondition(deployment.GetConditions(), condition)
	}
	requeue := !allConditionsComplete

	return requeue, nil
```

Comment (kylewuolle) [reply internal/controller/deployment_controller.go:115]: Kind of works thanks! Turns out the conditions field is a list of interfaces and not maps though so I had to change it a bit.

File: internal/controller/deployment_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kylewuolle) [reply internal/controller/deployment_controller.go]: Good point I've added that in to all of them.

File: internal/controller/deployment_controller.go (hunk 1 above)
Comment (kylewuolle) [reply internal/controller/deployment_controller.go:104]: Is the concern around uniqueness? I believe the cluster name must be unique within a namespace.

File: config/dev/deployment.yaml [hunk 5]
```diff
@@ -1,17 +1,17 @@
 apiVersion: hmc.mirantis.com/v1alpha1
```
Comment (squizzi) [inline config/dev/deployment.yaml:1]: Looks like you checked your `deployment.yaml` in, run a `git checkout upstream/main -- config/dev/deployment.yaml` and grab the original.

File: internal/controller/deployment_controller.go [hunk 6]
```diff
@@ -87,7 +89,66 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+func (r *DeploymentReconciler) setStatusFromClusterStatus(ctx context.Context, dc *dynamic.DynamicClient, l logr.Logger, deployment *hmc.Deployment) (bool, error) {
+	resourceId := schema.GroupVersionResource{
+		Group:    "cluster.x-k8s.io",
+		Version:  "v1beta1",
+		Resource: "clusters",
+	}
+
+	list, err := dc.Resource(resourceId).Namespace(deployment.Namespace).Get(ctx, deployment.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		l.Info("Cluster not found, ignoring since object must be deleted or not yet created")
```
Comment (squizzi) [inline internal/controller/deployment_controller.go:101]: This is a list object but the info returned implies a single cluster, perhaps `No clusters found` is a better log message here.

File: internal/controller/deployment_controller.go [hunk 7]
```diff
@@ -224,6 +285,20 @@ func (r *DeploymentReconciler) Update(ctx context.Context, l logr.Logger, deploy
...
+		if err != nil {
+			if requeue {
+				return ctrl.Result{RequeueAfter: 10 * time.Second}, err
+			} else {
+				return ctrl.Result{}, err
+			}
+		}
+
+		if requeue {
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
```
Comment (squizzi) [inline internal/controller/deployment_controller.go:299]: Instead of requeuing every 10 seconds, perhaps its better to add a `Watches` for the cluster object (once it's added to scheme) within `SetupWithManager`, that way when the status of the cluster object changes we would requeue rather then relying on polling.

File: internal/controller/deployment_controller.go (hunk 1 above)
Comment (Kshatrix) [reply internal/controller/deployment_controller.go:99]: No guarantee that cluster.Name == deployment.Name

Comment (kylewuolle) [reply internal/controller/deployment_controller.go]: Ah ok I've changed it to use the label

File: internal/controller/deployment_controller.go (hunk 7 above)
Comment (kylewuolle) [reply internal/controller/deployment_controller.go:304]: Looks like there's only one watch allowed per controller unfortunately.

File: internal/controller/deployment_controller.go [hunk 8]
```diff
@@ -87,7 +91,69 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	return r.Update(ctx, l, deployment)
 }
 
+func (r *DeploymentReconciler) setStatusFromClusterStatus(ctx context.Context, dc *dynamic.DynamicClient, l logr.Logger, deployment *hmc.Deployment) (bool, error) {
+	resourceId := schema.GroupVersionResource{
+		Group:    "cluster.x-k8s.io",
+		Version:  "v1beta1",
+		Resource: "clusters",
```
Comment (squizzi) [inline internal/controller/deployment_controller.go:98]: I don't want to scope creep you, perhaps it's in a different issue, but Is `clusters` the only resource that this status should be referencing?  What about `Workers` and specifically `MachineDeployment`'s, the cluster could be `ready: True` when the workers aren't yet up.

Comment (kylewuolle) [reply internal/controller/deployment_controller.go:98]: Michael said to just go with the clusters for now.

**Review by squizzi — CHANGES_REQUESTED** [review]

**Comment by squizzi** [comment]
@kylewuolle just need to rebase to get rid of this merge commit: https://github.com/Mirantis/hmc/pull/230/commits/0635d504fd94874418e1cf426e00c9e425a0dae3 and it lgtm

File: cmd/main.go [hunk 9]
```diff
@@ -163,6 +170,7 @@ func main() {
 		Client: mgr.GetClient(),
 		Scheme: mgr.GetScheme(),
 		Config: mgr.GetConfig(),
+		Dc:     dc,
```
Comment (squizzi) [inline cmd/main.go:173]: nit: Any specific reason for leaving this `Dc`, if I was in my IDE and typing out `r.` and looking at completions `DynamicClient` would be a much more descriptive "oh I need this" then `Dc`.

**Review by squizzi — APPROVED** [review]
